### DISCUSSION TITLE
LDM Notes for September 25th, 2023

### DISCUSSION THREAD

**Discussion by 333fred** [dotnet/csharplang#7559] [LDM Notes]
https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-09-25.md

## Agenda

- Primary constructors
- Defining well-defined behavior for collection expression types

**Comment by TonyValenti** [comment]
How does the compiler know if an enumerable is side-effecting?
RE Well-behaved collections:

I have a ```GetConsumingEnumerable``` extension methods that remove items from their underlying collections.  How do I mark this as side effecting?

**Reply by BreyerW** [reply]
You dont and compiler will not attempt to guardrail you. It simply assumes ALL `IEnumerable`s that you use with collection expressions are all well-behaving and have no side effects at all. From what i read on discord, team is overalll fine with this potential bug farm and handwaved the whole issue

Personally im not thrilled about this even if ends up not affecting me or very little but due to lack of determinism i cant be certain about this

**Reply by HaloFour** [reply]
The compiler won't know, it will assume that the enumerable doesn't have side-effects that would effect enumeration of any other spread enumerable.

It would be nice to get some kind of "guarantees" as leaving it open-ended definitely gives me pause and led to a bit of discussion on Discord yesterday.  The behavior sounds like it will be more intuitive than it sounds, but that the expression reserves the right to handle the enumerables in a way to optimize for performance and space.

So given the following:

```cs
IEnumerable<T> M() { ... }
IEnumerable<T> N() { ... }

var x = [ .. M(), .. N() ];
```

What you can expect is that `M()` will be invoked before `N()` is invoked.  What you can't expect is that the result from `M()` will be fully enumerated before `N()` is invoked.  It sounds like you also can't necessarily expect the result from `M()` to be fully enumerated before the result from `N()` is fully enumerated.  It sounds like you can expect that each enumerable will be enumerated only once, though.

So where this seemingly becomes a problem is if you have two enumerables where the latter enumerable depends on the first enumerable to be enumerated fully first.

It's possible that I've misinterpreted some of the conversation yesterday, and not everyone agreed with the behavior in the conclusion of this meeting.  So more feedback will likely be helpful here.  Undefined behavior today does not mean that the compiler can't become more prescriptive in a later release either.

**Reply by jnm2** [reply]
⬆️ This is a good summary.

**Reply by MrJul** [reply]
@HaloFour wrote:

> It sounds like you can expect that each enumerable will be enumerated only once, though.

It would be nice if this could be part of the specification if this was mentioned on discord, because right now the notes specify:
> We support spreading IEnumerables, not IEnumerators, and those should be iterable multiple times without producing different results each time.

Implying that the compiler is free to iterate the same `IEnumerable` several times. This is personally something I try to usually avoid in my code with arbitrary enumerables as there's no guarantee that enumerating the same enumerable twice will return the same values, or even if they do, the enumeration itself might be costly.

One doesn't have to search far to find examples: `Directory.EnumerateFiles` and most `IEnumerable<T>` backed by database `IQueryable<T>` have no guarantees to be trivially re-iterable.

**Reply by jnm2** [reply]
@333fred Could you clarify the notes? They do read as if to say that a given spread may be enumerated multiple times.

**Reply by CyrusNajmabadi** [reply]
My recollection from the meeting was that we were saying values should behave reasonably and responsibly. Similar to the rules we have for patterns. So evaluating/iterating one thing should not cause observable changes in other parts of the collection expression (outside of LTR evaluation guarantees). And collections should assume they may be introspected (both statically and dynamically at runtime) so they compiler can make better decisions about what to do. Collections should not be written in a way that cares if that happens. If they do care,  we'd consider that the strange/out-of-norm case, and would be fine saying that collection expressions don't design for that case, but for the 99.9% case of normal collections where that is not a concern.

**Reply by jnm2** [reply]
But I emphasize that the things we spread may not _be_ collections. They may be other types of enumerables, such as queryables. A well-behaved enumerable may return a completely difference sequence each time it is enumerated.

**Reply by jnm2** [reply]
I would go so far as to say that the things we spread should not be thought of as collections unless they implement one of the collection interfaces above and beyond IEnumerable. Even the presence of Count/Length properties on an enumerable isn't quite a statement that "I am a collection."

**Reply by CyrusNajmabadi** [reply]
We don't have such rules for list patterns. We make a similar assumption from the presence of certain members. I'm perfectly ok with that. If that proves to not be accurate for a particular API, I'm ok with that. For example, if you are iterable, and have a Count, and that doesn't behave as one would expect *I don't have a problem*. 

Such an API feels already confusing and problematic to me.  And I think such an API should simply ship with an analyzer saying "don't use me here, I won't work as one might expect" :-)

**Reply by HaloFour** [reply]
Collection introspection aside, is there a reason for spreads or list patterns to need to enumerate an `IEnumerable<T>` more than once?  Would it be problematic to specify that the language features would ensure that such an enumerable should only be enumerated at most once?

**Reply by CyrusNajmabadi** [reply]
I don't like a strict requirement as it means we cannot call any helpers as we would not be able to guarantee anything. Even calling a property might cause enumeration, and that would things get limiting. I'd simply prefer to state that elements of collections be well behaved across the patterns we might use. 

This is the same thing we did for list patterns (and patterns in general), and it's not proven itself to be a problem (even though you could totally have a strangely behaving type where you observed weirdness). 

I think we simply want to give us the flexibility to be good for all expected cases, even if that means you don't get guarantees for strange cases. That is a very sensible place for me, and lands in the current "well behaved" approach we've been designing by.

**Reply by HaloFour** [reply]
> I'd simply prefer to state that elements of collections be well behaved across the patterns we might use.

Right, but to @jnm2 's point, not all `IEnumerable<T>`s are collections, and I could definitely see people wanting to use list literals with LINQ and queryables in order to combine/assemble results.  Accidentally making duplicate DB/network calls to do so would be unexpected and undesirable.  To that end, I don't think it's worthwhile for the compiler to make assumptions as to how bare enumerables would work, at least as far as it being safe to enumerate them multiple times without some kind of memoization.

As for enumerables that pretend to be collections by adhering to the patterns of exposing `Count`/`Length` properties or indexers, sure, all bets are off.

**Reply by CyrusNajmabadi** [reply]
>As for enumerables that pretend to be collections by adhering to the patterns of exposing Count/Length properties or indexers, sure, all bets are off.

ok.  but then i don't get how to reasonably prevent one without the other.  'Count' could be a property exposed by a collection that actually does the enumeration.  The only way we could actually spec that it happens once is to state that we won't do *anything* *at all* except the enumeration.  Which leaves us unable to optimize the 99.99% case.

I woudl rather us say: we make presumptions that your type is not insane, and that both the members of the type we call, as well as helpers we look for, will all behave in a *sane*, *well behaved* fashion.  And, if that's not the case, *c'est la vie*.

> To that end, I don't think it's worthwhile for the compiler to make assumptions as to how bare enumerables would work

I strongly disagree.  Note: Linq itself *disagrees*.  It makes assumptions about bare enumerables all the time.  Presuming that just the presence of certain things it can reflect on should impact performance.

Note, it has even added this stuff over the years, *breaking* programs that depended on the old behavior.  The literal runtime did this, even with breaks, because the gain the ecosystem *overall* was far too great.  It makes no sense to me that we have that experience, but would now be more conservative at the *language* level.

**Reply by HaloFour** [reply]
I think we're largely on the same page.  If the enumerable smells like a collection, then the compiler can make some degree of assumptions as to how it behaves under the hood.  The compiler shouldn't need to guard against pathological cases.

I do think we need to be a little wary on how that may impact a future with support for extension properties, in that I think the docs should be really clear that adding an extension `Count` property could result in a "Bad Time" ™️.

> I strongly disagree. Note: Linq itself _disagrees_. It makes assumptions about bare enumerables all the time. Presuming that just the presence of certain things it can reflect on should impact performance.

I don't believe I've run into a LINQ method that will result in multiple enumerations of the input enumerables?  Even with `Join` and the like my understanding is that they enumerate once and materialize the results into another collection.

Either way, I think it's reasonable to assume that spread expressions within a list expression will be used to flatten queryables, and it shouldn't be a pit of failure.

**Reply by CyrusNajmabadi** [reply]
> in that I think the docs should be really clear that adding an extension Count property could result in a "Bad Time" ™️.

Right.  Note that we've crossed that bridge already with patterns.  And we have effectively gone: if you do bad things, you get bad results.   Note: i will *expect* taht when we have 'extension everything' that a type could become list-patternable just by exposing the right extension.  Our presumption on the compiler side is htat if you expose that you are being sane, and not doing pathological things.  If that's not the case, we won't shed a tear :)

Our position is that it's better *overall* to assume well-behavedness, and allow those assumptions to end up majorly benefitting the 99.99% of people not doing bizarre stuff, than to be too worried about people pulling "WATs" and then ruining it for everyone else :)

>  and it shouldn't be a pit of failure.

I would not expect any pit of failure here.  ANd i think we'd discover it right away if it was.

**Reply by jnm2** [reply]
@CyrusNajmabadi Sure, instead of requiring a collection interface to be implemented, you could just look for Count/Length like patterns do. But if there is no Count/Length, this thing you're spreading might not be a collection and should not be expected to behave like a well-behaved collection. A well-behaved enumerable might return different results each time and might have important side effects (like expensive queries).

**Reply by CyrusNajmabadi** [reply]
I'm fine saying: if you use it in a collection expression, it should be well behaved.  Similar to how we say the same for patterns.

**Reply by jnm2** [reply]
I think we're talking past each other. There are two kinds of well-behavedness that should not be confused: a well-behaved enumerable, versus a well-behaved collection. A well-behaved enumerable should **not** be expected to return the same sequence if enumerated multiple times. The current notes are confusing on this point.

**If** the enumerable happens to be collection-like (defining Count), then it becomes a _collection_ and can be expected to be a well-behaved _collection_ which should enumerate the same items multiple times without side effects.

If not, the thing being spread is not a collection but is some other type of well-behaved enumerable, and thus the compiler simply cannot enumerate it multiple times. That's not a general expectation of a well-behaved enumerable.

**Reply by sab39** [reply]
Strongly agree that the spec should guarantee that a spread element will only be (directly) enumerated once. It's an explicit goal of collection expressions to perform at least as well as manually written code, and even if we _do_ define "well behaved enumerable" to require giving the same result from repeated enumerations, it's _certainly_ the case that enumeration can be expensive. It would be highly surprising for a collection expression to perform substantially worse than manually written code for such an enumerable.

Note that I said "directly" enumerated: of course the compiler can't guarantee that a `Count` property won't also cause enumeration as a side effect, but the compiler can and should guarantee that the code _it_ generates will call `GetEnumerator()` at most once for each spread element.

(Sidenote: has consideration been given to types like linked lists, which might implement `IList<T>`, can be enumerated efficiently, but are O(n) expensive to index into? Should the compiler guarantee that it will use an enumerator rather than calling index accessors in some circumstances to avoid O(n^2) complexity? If so, which circumstances?)

**Reply by CyrusNajmabadi** [reply]
> Strongly agree that the spec should guarantee that a spread element will only be (directly) enumerated once. 

> Note that I said "directly" enumerated: of course the compiler can't guarantee that a Count property won't also cause enumeration as a side effect

If we cannot actually guarantee anything, i don't see the purpose of putting in pretend guarantees.  Better to be honest that we simply do not have this guarantee.  In other words, *no promise* is better than a *false promise*.  

> Should the compiler guarantee

Compiler guarantees can be elsewhere.  I have no issue with the compiler writing up a list of rules/heuristics/guardrails, it works within.  But as far as the language is concerned.  No.  The language should either allow the code to be written or not.  It's then up to the compiler ot make intelligent rules that benefit the vast majority, do not punish reasonable cases, but also potentially disregard the pathlogical.

**Reply by CyrusNajmabadi** [reply]
Note: We are planning to discuss this (and if spreads give LTR ordering wrt to their own independent enumeration) to see if we want to give any additional rules/guidance.

**Reply by En3Tho** [reply]
Consider this very silly example (I guess it can be called "insane" in context of this thread):
```csharp
public class SillyCollection<T>(IEnumerable<T> Enumerable) : ICollection<T>
{
    public IEnumerator<T> GetEnumerator() => Enumerable.GetEnumerator();    
    public int Count => Enumerable.Count();
   // other stuff ...
}
```

If compiler implements any optimizations like for example `TryGetNonEnumeratedCount()` it automatically will enumerate the thing twice. I woudn't personally expect compiler to call `.GetEnumerator()` more that once for a given enumerable but giving a strong guarantee about enumerating thing only once is too limiting if not simply wrong.

**Reply by sab39** [reply]
> If we cannot actually guarantee anything, i don't see the purpose of putting in pretend guarantees. Better to be honest that we simply do not have this guarantee. In other words, _no promise_ is better than a _false promise_.

But we _can_ guarantee something: we can guarantee that the compiler will not insert more than one enumeration of the collection. The compiler can't know that `Count` or some other property won't cause an enumeration as a side effect, but the _programmer_ can know that about the type they're using.

Consider the _extremely_ common case of enumerables created by generator methods - which I would expect account for the vast majority of non-collection enumerables. In that case it's inherently impossible for the enumerable to come with a `Count` property because it's guaranteed that the code of the generator won't execute until the result is being enumerated. I would personally find it not just unexpected but absolutely astonishing if the code of my generator method were executed more than once when I used it a single time in a collection expression.

Worse: if I _do_ need a guarantee that my generator method code will only be run once, and the compiler doesn't give such a guarantee, then my only option is to insert a `ToList()` call or similar, which then allocates unnecessary storage and results in redundant copying. Or to use something other than collection expressions.

Worse still: if the language _reserves the right_ for the compiler to enumerate spread elements more than once but doesn't actually do so in the initial release, it's all but certain that code will be written inadvertently relying on this fact. Which would then make it a breaking change in practice anyway if the compiler ever started doing it, even though the code was technically relying on undefined behavior in the language.

**Reply by CyrusNajmabadi** [reply]
@sab39 we're going to bring this up for discussion.  

> Which would then make it a breaking change in practice anyway

That's not the case.  That's the point about explicitly not making promises.  :)

Note: this is not hypothetical, around things like well-behavedness, we will not consider it break, since it's depending on explicitly undefined behavior.

That said, we are going to bring this up.  I'm not opposed to some additional statements here if it builds some confidence for these esoteric cases.
